### PR TITLE
Pass RepeatedTask to its closure #514

### PR DESCRIPTION
**Pull request by FranzBusch**
Pass the reference to `RepeatedTask` to its closure to allow a thread-safe cancelling. https://github.com/apple/swift-nio/issues/514

### Motivation:

With the current implementation, one has to create a reference and use a lock to thread-safely cancel the `RepeatedTask`. This makes the use more difficult and not so straight-forward.

### Modifications:

Added a weak `RepeatedTask?` parameter to the `scheduleRepeatedTask` closure.

### Result:

It is now possible to cancel the `RepeatedTask` from within the task closure.

### CODE REVIEW

**Comment by FranzBusch** [comment]
@Lukasa Can you check if this usage of the weak reference is fine? I used since I couldn't figure out any other way not to create a retain cycle and breaking the test `testScheduleRepeatedTaskToNotRetainEventLoop`.

File: Sources/NIO/EventLoop.swift [hunk 1]
```diff
@@ -59,10 +59,12 @@ public final class RepeatedTask {
     private var scheduled: Scheduled<EventLoopFuture<Void>>?
     private var task: (() -> EventLoopFuture<Void>)?
 
-    internal init(interval: TimeAmount, eventLoop: EventLoop, task: @escaping () -> EventLoopFuture<Void>) {
+    internal init(interval: TimeAmount, eventLoop: EventLoop, task: @escaping (RepeatedTask?) -> EventLoopFuture<Void>) {
         self.delay = interval
         self.eventLoop = eventLoop
-        self.task = task
+        self.task = { [weak self] in
```
Comment (Lukasa) [inline Sources/NIO/EventLoop.swift]: You can avoid the weak reference by recreating the task each time. That is, you call `scheduleTask(in: delay) { task(self) }`.

That does provide a reference cycle, but it's one that already exists, as `reschedule` already gives the `Future` a reference to `self`. Then there's no need to receive the `RepeatedTask` as an optional.

File: Sources/NIO/EventLoop.swift [hunk 2]
```diff
@@ -116,14 +120,20 @@ public final class RepeatedTask {
                 self.reschedule0()
             }
         }
+
+        scheduled.futureResult.whenFailure { _ in
```
Comment (Lukasa) [inline Sources/NIO/EventLoop.swift]: Can you ignore the result but specify its type? That helps us catch unexpected API changes. E.g., write `{ (_: Error) in`

File: Tests/NIOTests/EventLoopTest.swift [hunk 3]
```diff
@@ -144,7 +131,7 @@ public class EventLoopTest : XCTestCase {
         let expect = expectation(description: "Is cancelling RepatedTask")
         let group = DispatchGroup()
         group.enter()
-        let repeatedTask = eventLoopGroup.next().scheduleRepeatedTask(initialDelay: initialDelay, delay: delay) { () -> Void in
+        let repeatedTask = eventLoopGroup.next().scheduleRepeatedTask(initialDelay: initialDelay, delay: delay) { _ -> Void in
```
Comment (Lukasa) [inline Tests/NIOTests/EventLoopTest.swift]: Can you ignore this value but specify its type, please? Same with the two changes below.

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Pull request merged**